Closes issue #7762: Adds support for persisting/restoring downloads.

Amejia481 opened the pull request:
Fenix pr https://github.com/mozilla-mobile/fenix/pull/14043
---
<!-- Text above this line will be added to the commit once "bors" merges this PR -->

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

csadilek commented on components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlinx/coroutines/Utils.kt:
```diff
@@ -0,0 +1,37 @@
+ *
+ * Credit to Terenfear https://gist.github.com/Terenfear/a84863be501d3399889455f391eeefe5
+ *
+ * @param skipTime the time to wait until the next call to [block] be processed.
```
Let's make it clear that this is in ms, maybe even by naming the param `skipTimeInMs`.

csadilek replied:
Another good thing to point out in the docs, is that the calls to the returned function should happen on the same thread, esp. for `throttleLatest` (see comment below). Unless we make sure we have visibility guarantees. Not a problem the way you use it now though.

csadilek commented on components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlinx/coroutines/Utils.kt:
```diff
@@ -0,0 +1,37 @@
+ * @param coroutineScope the coroutine scope where [block] will executed.
+ * @param block function to be execute.
+ */
+fun <T> throttle(
```
Good find. I think you want to use the `throttleLatest` from the example above though? Then after skipTime you would get the latest/newest update to the store. Now you get the oldest.

Amejia481 replied:
Thanks!

csadilek commented on components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt:
```diff
@@ -590,6 +590,11 @@ sealed class DownloadAction : BrowserAction() {
+    /**
+     * Restore the [BrowserState.downloads] state from the storage.
+     */
+    object RestoreDownloadsState : DownloadAction()
```
nit: Let's keep the `Action` postfix. Eventually, I think we can remove `Action` from all these names. I filed a ticket for that a while ago, but for now let's keep it consistent.

csadilek commented on components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadMiddleware.kt:
```diff
@@ -7,34 +7,100 @@ package mozilla.components.feature.downloads
-                intent.putExtra(DownloadManager.EXTRA_DOWNLOAD_ID, action.download.id)
-                applicationContext.startService(intent)
+                next(action)
+                scope.launch {
```
Here we should check before we launch. No need to launch the coroutine if we don't need to add to the storage.

csadilek commented on components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt:
```diff
@@ -574,12 +574,12 @@ sealed class DownloadAction : BrowserAction() {
      * Updates the [BrowserState] to track the provided [download] as added.
      */
-    data class AddDownloadAction(val download: DownloadState) : DownloadAction()
+    data class AddDownloadAction(val download: DownloadState, val restored: Boolean = false) : DownloadAction()
```
Could `restored` just be part of the `DownloadState` instead of being in the action?

Amejia481 replied:
Do you think that we are going to need it in the future?

Amejia481 replied:
This is the only item left on the updated. I'm not sure if we are going to need `restored`  on `DownloadState` in the future, as now it just serves the purpose of identifying a download is coming from the storage.

Amejia481 replied:
👍

Amejia481 replied:
Added a new `RestoreDownloadAction` makes things clear :)

csadilek commented on components/browser/state/src/main/java/mozilla/components/browser/state/state/content/DownloadState.kt:
```diff
@@ -40,40 +40,42 @@ data class DownloadState(
      * Status that represents every state that a download can be in.
      */
-    enum class Status {
+    @Suppress("MagicNumber")
```
Couldn't we just persist the String value and rely on `Enum.valueOf`. Then we don't need the `StatusConverter` etc.

Amejia481 replied:
It could be a bit brittle if somebody renames some `Status`, at least with `ids` is less possible, what do you think?

csadilek replied:
Well, on the flip side, if someone renames / messes up the names we would get an exception/crash. If someone messes up IDs crazy things can start happening. Eitherway, we need data migration if the status fields change. Up to you :). Don't feel strongly, just thought we can simplify.

Amejia481 replied:
👍

Amejia481 replied:
For the moment, I prefer to keep it as it's :)

csadilek commented on components/feature/downloads/schemas/mozilla.components.feature.downloads.db.DownloadsDatabase/1.json:
```diff
@@ -0,0 +1,76 @@
+          {
+            "fieldPath": "status",
+            "columnName": "status",
+            "affinity": "INTEGER",
```
See my comment / question above. :)

csadilek commented on components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadMiddleware.kt:
```diff
@@ -7,34 +7,100 @@ package mozilla.components.feature.downloads
+                        downloads.forEach { download ->
+                            if (!store.state.downloads.containsKey(download.id)) {
+                                store.dispatch(DownloadAction.AddDownloadAction(download, true))
+                                logger.error("Download restarted from db")
```
This shouldn't be logged as `error` and should probably be restored, not restarted?

Amejia481 replied:
Agree, It's a leftover from testing.

csadilek commented on components/feature/downloads/src/main/java/mozilla/components/feature/downloads/db/DownloadsDatabase.kt:
```diff
@@ -0,0 +1,55 @@
+}
+
+@Suppress("unused")
+internal class StatusConverter {
```
See comment above. I think I would prefer not needing this, but I might miss why this is more practical?

Amejia481 commented:
@csadilek I addressed the comments on this commit https://github.com/mozilla-mobile/android-components/pull/8197/commits/1c374b005c647975d683faab19c32fd1fe24fdd1 :)

csadilek commented on components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadStorage.kt:
```diff
@@ -0,0 +1,89 @@
+         * only takes into considerations fields that are being stored,
+         * not all the field on [DownloadState] are stored.
+         */
+        fun areTheSame(first: DownloadState, second: DownloadState): Boolean {
```
I think we should move or rename this method. `DownloadStorage.areTheSame(a, b)` doesn't look right. It doesn't check if tow storages are the same, but compares `DownloadState`s. I would move to `DownloadState` or turn into an ext. function.

Amejia481 replied:
I didn't want to added directly to `DownloadState` as it could get confused with `equals`, because It only cares about properties that the storage take into consideration, we don't want to overwhelm the storage with every update on the `DownloadMiddleware`  if nothing that the storage cares has changed  : 
```
            is DownloadAction.UpdateDownloadAction -> {
                val updated = action.download
                store.state.downloads[updated.id]?.let { old ->
                    // To not overwhelm the storage, we only send updates that are relevant,
                    // we only care about properties, that we are stored on the storage.
                    if (!DownloadStorage.areTheSame(old, updated)) {
                        scope.launch {
                            downloadStorage.update(action.download)
                        }
                        logger.debug("Updated download ${action.download.fileName} on the storage")
                    }
                }
            }
```

csadilek replied:
Sure, but I don't think how it would be confused with `equals`, but you can simply rename the method too and keep it in the storage class. My comment was just about `areTheSame` not being a good name on the storage. When I read the tests I found this very confusing. `compareDownloads` etc. would have been much cleared, imo.

csadilek commented on components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadStorage.kt:
```diff
@@ -0,0 +1,89 @@
+         * not all the field on [DownloadState] are stored.
+         */
+        fun areTheSame(first: DownloadState, second: DownloadState): Boolean {
+            return first.id == second.id &&
```
Now that we have a unique ID, comparing the ID should be good enough here. This would hide a bug if we had two downloads with the same ID but different file name.

Amejia481 replied:
We are using this function to validate if something has changed, between updates. We need to verify for each field that the storage cares about, because we don't overwhelm the storage, if there is not new info, maybe `DownloadStorage` is not a good place. What do you think?


```kotlin
            is DownloadAction.UpdateDownloadAction -> {
                val updated = action.download
                store.state.downloads[updated.id]?.let { old ->
                    // To not overwhelm the storage, we only send updates that are relevant,
                    // we only care about properties, that we are stored on the storage.
                    if (!DownloadStorage.areTheSame(old, updated)) {
                        scope.launch {
                            downloadStorage.update(action.download)
                        }
                        logger.debug("Updated download ${action.download.fileName} on the storage")
                    }
                }
            }
```

csadilek replied:
OK, see comment below. I think we can just choose a clearer name and keep everything as-is otherwise.

Amejia481 replied:
Any suggestions for the new name?

Amejia481 replied:
What about `DownloadStorage.equals(a, b)` ?

csadilek commented on components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlinx/coroutines/Utils.kt:
```diff
@@ -0,0 +1,41 @@
+ *
+ * Returns a function that limits the executions of the [block] function, until the [skipTimeInMs]
+ * passes, then the latest value passed to [block] will be used. Any calls before [skipTimeInMs]
+ * passes will be ignored. It's spected that calls to the returned function
```
nit: "expected"? I would write: All calls to the returned function must happen on the same thread.

Amejia481 replied:
👍

csadilek commented on components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadMiddleware.kt:
```diff
@@ -7,34 +7,100 @@ package mozilla.components.feature.downloads
+
+    @InternalCoroutinesApi
     override fun invoke(
         store: MiddlewareStore<BrowserState, BrowserAction>,
```
We have an API change here on master. So we need to update this to receive a `MiddlewareContext` (or simply rebase). Everything else will be fine e.g. you can still `dispatch` on it etc.

Amejia481 replied:
👍

Amejia481 commented:
New changes are in https://github.com/mozilla-mobile/android-components/pull/8197/commits/c1a18b11dc38cde3d77e7e9f6183b58fc5604455

csadilek commented on components/browser/state/src/main/java/mozilla/components/browser/state/reducer/DownloadStateReducer.kt:
```diff
@@ -25,6 +25,8 @@ internal object DownloadStateReducer {
             is DownloadAction.RemoveAllDownloadsAction -> {
                 state.copy(downloads = emptyMap())
             }
+            DownloadAction.RestoreDownloadsStateAction -> state
```
nit: no `is`

We should also add a quick test case that this action doesn't affect the state :) e.g

```
    @Test
    fun `RestoreDownloadsStateAction does nothing`() {
        val store = BrowserStore(BrowserState())

        val state = store.state
        store.dispatch(DownloadAction.RestoreDownloadsStateAction).joinBlocking()
        assertSame(store.state, state)
    }
```

Amejia481 replied:
👍

csadilek commented on components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadMiddleware.kt:
```diff
@@ -7,35 +7,116 @@ package mozilla.components.feature.downloads
+                sendDownloadIntent(action.download)
+            }
+        }
+        if (action !is DownloadAction.AddDownloadAction) {
```
That's a bug here now. `RestoreDownloadStateAction` would cause next to be called twice. I would handle this differently and have a `preProcess` function that returns whether nor not `next` should be called e.g. whether or not the original action should hit the store.

Here's an example:  https://github.com/mozilla-mobile/android-components/blob/master/components/feature/readerview/src/main/java/mozilla/components/feature/readerview/ReaderViewMiddleware.kt#L55

Amejia481 replied:
Thanks for catching this 👍

Amejia481 replied:
What do you think about 
```kotlin
        if (requireNext(context, action)) {
            next(action)
            postNextProcess(action)
        }
```

csadilek replied:
I find that a bit confusing, to be honest :).

csadilek replied:
So, I would check actions individually or do what we did in the `ReaderViewMiddleware` so we follow the same pattern.

Amejia481 commented:
Pr update https://github.com/mozilla-mobile/android-components/pull/8197/commits/3b9703e1b68f82bd3357e631cf6e51ac6c73a538 :)

csadilek commented on components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:
```diff
@@ -625,7 +630,15 @@ abstract class AbstractFetchDownloadService : Service() {
-                " currentBytesCopied ${downloadJobState.currentBytesCopied}")
+                " currentBytesCopied ${downloadJobState.state.currentBytesCopied}")
+
+        val throttleUpdatedDownload = throttleLatest<Long>(
```
nit: `throttleUpdateDownload`

csadilek commented on samples/browser/src/main/java/org/mozilla/samples/browser/SampleApplication.kt:
```diff
@@ -58,7 +59,7 @@ class SampleApplication : Application() {
             components.webAppManifestStorage.warmUpScopes(System.currentTimeMillis())
         }
-
+        components.store.dispatch(DownloadAction.RestoreDownloadsStateAction)
```
nit: Was wondering if we should create a use case to `DownloadsUseCases` for this.

Amejia481 replied:
Agree, it will hide the implementation details!

csadilek commented on components/feature/downloads/src/test/java/mozilla/components/feature/downloads/DownloadMiddlewareTest.kt:
```diff
@@ -8,34 +8,181 @@ import android.app.DownloadManager.EXTRA_DOWNLOAD_ID
+
+        verify(downloadStorage).update(updatedDownload)
+
+        updatedDownload = download.copy(url = "updatedURLAgain")
```
nit: use a test case that can happen...this seems a bit contrived?

Amejia481 replied:
updated!

csadilek commented on components/feature/downloads/src/test/java/mozilla/components/feature/downloads/DownloadMiddlewareTest.kt:
```diff
@@ -8,34 +8,181 @@ import android.app.DownloadManager.EXTRA_DOWNLOAD_ID
+    }
+
+    @Test
+    fun `UpdateDownloadAction MUST update the storage when changes are meaningful`() = runBlockingTest {
```
nit: when changes are needed :)

Amejia481 replied:
👍

Amejia481 commented:
bors r=csadilek
